### PR TITLE
Chapter 11: NgRx & Chapter 12: Material Components

### PR DESCRIPTION
**Pull request by schintees**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -21,6 +21,10 @@
     "@angular/platform-browser-dynamic": "^16.1.0",
     "@angular/router": "^16.1.0",
     "@ngneat/until-destroy": "^10.0.0-beta.0",
+    "@ngrx/effects": "^16.1.0",
+    "@ngrx/store": "^16.1.0",
+    "@ngrx/store-devtools": "^16.1.0",
+    "ngrx-store-localstorage": "^16.0.0",
```
Comment (antonio-marasescu-duran) [inline package.json]: **question:* why did you add this?

File: src/app/app.module.ts [hunk 2]
```diff
@@ -57,6 +64,11 @@ import { UserModule } from './modules/user/user.module';
         ReactiveFormsModule,
         MatInputModule,
         MatCardModule,
+        MatProgressSpinnerModule,
+        MatTableModule,
+        StoreModule.forRoot(appReducers, appMetaReducers),
+        EffectsModule.forRoot(appEffects),
+        StoreDevtoolsModule.instrument(),
```
Comment (antonio-marasescu-duran) [inline src/app/app.module.ts]: **suggestion:** is best to not have the store devtools available in production, use the `environment.production` variable here

File: package.json (hunk 1 above)
Comment (schintees) [reply package.json]: I added it to synchronize part of the state with localstorage, to maintain the state in case of refresh

File: src/app/components/containers/products-details/products-details.component.ts [hunk 3]
```diff
@@ -2,44 +2,43 @@ import { Component, OnInit } from '@angular/core';
...
   templateUrl: './products-details.component.html',
   styleUrls: []
 })
 export class ProductsDetailsComponent implements OnInit {
-  product$?: Observable<Product>;
-  isAdmin$?: Observable<boolean>;
-  isCustomer$?: Observable<boolean>;
-  
+  product$: Observable<Product | undefined> = this.store.select(selectProduct);
+  isAdmin$: Observable<boolean | undefined> = this.store.select(selectIsAdmin);
```
Comment (antonio-marasescu-duran) [inline src/app/components/containers/products-details/products-details.component.ts]: **suggestion(non-blocking):** Just as a note, I prefer keeping the select in ngOnInit as way to separate the notion of class instantiation over the the actual component instantiation.

File: src/app/modules/shopping-cart/state/cart.reducers.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        products: state.products.filter(cartProduct => cartProduct.product.id !== productId)
+    })),
+
+    // Add Product To Cart
+    on(addProductToCart, (state, { product }) => {
+        const existingCartProduct = state.products.find(cartProduct => cartProduct.product.id === product.id);
+        return {
+            ...state,
+            products:
+                existingCartProduct ?
```
Comment (antonio-marasescu-duran) [inline src/app/modules/shopping-cart/state/cart.reducers.ts]: **suggestion(non-blocking):** just as general rule of thumb try to avoid complex login inside the reducer handler (this is usually kept in the effects or in a service)

Comment (antonio-marasescu-duran) [reply src/app/modules/shopping-cart/state/cart.reducers.ts]: Here is good enough, though you could have done everything with a `filter` followed by a `concat` and will have looked nicer and simpler

File: src/app/modules/user/state/user.selectors.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import { createSelector } from "@ngrx/store";
+import { AppState } from "src/app/state/app.state";
+import { UserState } from "./user.reducers";
+import { UserRoles } from "../types/roles.enum";
+
+export const selectUserState = (state: AppState) => state.user;
```
Comment (antonio-marasescu-duran) [inline src/app/modules/user/state/user.selectors.ts]: **suggestion(non-blocking):** usually selectors are kept with the reducer. Though if the reducer file grows to large state (either you split the reducer or move the selectors). But as a starting point is better and simpler to have them in one place at the start.

File: src/app/modules/user/state/user.selectors.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+);
+
+export const selectIsAdmin = createSelector(
+    selectUserState,
+    (state: UserState) => state.currentUser?.roles.includes(UserRoles.Admin)
+)
+
+export const selectIsCustomer = createSelector(
+    selectUserState,
+    (state: UserState) => state.currentUser?.roles.includes(UserRoles.Customer)
```
Comment (antonio-marasescu-duran) [inline src/app/modules/user/state/user.selectors.ts]: **suggestion:** take care when you use "?" you need to add it to the rest of the following properties you access

File: src/app/state/app.effects.ts [hunk 7]
```diff
@@ -0,0 +1,9 @@
+import { ShoppingCartEffects } from "../modules/shopping-cart/state/cart.effects";
+import { UserEffects } from "../modules/user/state/user.effects";
```
Comment (antonio-marasescu-duran) [inline src/app/state/app.effects.ts:2]: **suggestion(non-blocking):** since you started separating into modules even more than necessary, keep in mind that state that exists for a specific module, can be lazy loaded there using `StoreModule.forFeature(<feature_name>, <feature_reducer>` and `EffectsModule.forFeature([<list_of_effect_classes>])`.

Now since you already done it and it was outside of the initial scope, don't change it. But keep in mind how is properly done

File: package.json (hunk 1 above)
Comment (antonio-marasescu-duran) [reply package.json]: This was not exactly a requirement, and we usually don't want to cache the data without a clear Caching Strategy (when to invalidate it?)

**Review by antonio-marasescu-duran — APPROVED** [review]
